### PR TITLE
feat(e2ei): respect E2EI during login and client creation (WPB-5851)

### PR DESCRIPTION
**Pull request by mchenani**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?
Respecting E2EI state during MLS client registration and login.

Needs releases with:

- [] GitHub link to other pull request

### Testing
WIP
#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/RegisterMLSClientUseCase.kt [hunk 1]
```diff
@@ -18,32 +18,55 @@
...
+    override suspend operator fun invoke(clientId: ClientId): Either<CoreFailure, RegisterMLSClientResult> =
+        mlsClientProvider.getMLSClient(clientId).flatMap { mlsClient ->
+            userConfigRepository.getE2EISettings().fold({
+                Either.Right(mlsClient)
+            }, { e2eiSettings ->
+                if (e2eiSettings.isRequired && !mlsClient.isE2EIEnabled()) {
+                    kaliumLogger.i("MLS Client registration stopped: e2ei is required and is not enrolled!")
+                    return Either.Right(RegisterMLSClientResult.E2EICertificateRequired(mlsClient))
+                } else Either.Right(mlsClient)
+            })
```
Comment (borichellow) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/RegisterMLSClientUseCase.kt:65]: could be simpler 
```
        mlsClientProvider.getMLSClient(clientId).map { mlsClient ->
            val e2eiSettings = userConfigRepository.getE2EISettings().getOrNull()
            if (e2eiSettings?.isRequired == true && !mlsClient.isE2EIEnabled()) {
                RegisterMLSClientResult.E2EICertificateRequired(mlsClient)
            } else {
                mlsClient
            }
        }
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/RegisterMLSClientUseCase.kt [hunk 2]
```diff
@@ -18,32 +18,55 @@
...
+        }.flatMap {
+            clientRepository.registerMLSClient(clientId, it.getPublicKey())
+        }.flatMap {
+            keyPackageRepository.uploadNewKeyPackages(clientId, keyPackageLimitsProvider.refillAmount())
+            Either.Right(RegisterMLSClientResult.Success)
+        }
```
Comment (borichellow) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/RegisterMLSClientUseCase.kt:71]: simpler:
```
        }.map {
            keyPackageRepository.uploadNewKeyPackages(clientId, keyPackageLimitsProvider.refillAmount())
            RegisterMLSClientResult.Success
        }
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/e2ei/usecase/EnrollE2EIUseCase.kt [hunk 3]
```diff
@@ -41,19 +43,21 @@ interface EnrollE2EIUseCase {
...
         kaliumLogger.i("start E2EI Enrollment Initialization")
 
-        e2EIRepository.fetchTrustAnchors().onFailure {
-            return E2EIEnrollmentResult.Failed(E2EIEnrollmentResult.E2EIStep.TrustAnchors, it).toEitherLeft()
-        }
+        e2EIRepository.initE2EIClient(isNewClient = isNewClientRegistration)
+
+//         e2EIRepository.fetchTrustAnchors().onFailure {
+//             return E2EIEnrollmentResult.Failed(E2EIEnrollmentResult.E2EIStep.TrustAnchors, it).toEitherLeft()
+//         }
```
Comment (borichellow) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/e2ei/usecase/EnrollE2EIUseCase.kt:60]: is it commented temporary? Some `TODO` or explanation comment would be nice, if so

**Review by borichellow — APPROVED** [review]
just a few style suggestion, that are not mandatory for the feature

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/e2ei/usecase/EnrollE2EIUseCase.kt (hunk 3 above)
Comment (mchenani) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/e2ei/usecase/EnrollE2EIUseCase.kt:60]: fixed in other PR after this gets merged.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt [hunk 4]
```diff
@@ -1368,6 +1380,8 @@ class UserSessionScope internal constructor(
     val observeLegalHoldStateForUser: ObserveLegalHoldStateForUserUseCase
         get() = ObserveLegalHoldStateForUserUseCaseImpl(clientRepository)
 
+    suspend fun observeIfE2EIRequiredDuringLogin(): Flow<Boolean?> = clientRepository.observeIsClientRegistrationBlockedByE2EI()
+
```
Comment (ohassine) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt:1384]: Could we declare it as a use case here ?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/ObserveIsE2EIRequiredState.kt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.logic.feature.client
+
+import com.wire.kalium.logic.data.client.ClientRepository
+import kotlinx.coroutines.flow.Flow
+
+interface ObserveIsE2EIRequiredState {
```
Comment (ohassine) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/ObserveIsE2EIRequiredState.kt:23]:
```suggestion
interface ObserveIsE2EIRequiredStateUseCase {
```

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/FinalizeMLSClientAfterE2EIEnrollment.kt [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.logic.feature.client
+
+import com.wire.kalium.logic.data.client.ClientRepository
+import com.wire.kalium.logic.data.conversation.JoinExistingMLSConversationsUseCase
+import com.wire.kalium.logic.functional.map
+
+interface FinalizeMLSClientAfterE2EIEnrollment {
```
Comment (ohassine) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/FinalizeMLSClientAfterE2EIEnrollment.kt:24]:
```suggestion
interface FinalizeMLSClientAfterE2EIEnrollmentUseCase
```

File: testservice/src/main/kotlin/com/wire/kalium/testservice/managed/ConversationRepository.kt [hunk 7]
```diff
@@ -120,6 +124,67 @@ sealed class ConversationRepository {
                 }
             }
 
+        suspend fun sendButtonAction(
+            instance: Instance,
+            conversationId: ConversationId,
+            referenceMessageId: String,
+            buttonId: String
+        ): Response = instance.coreLogic.globalScope {
```
Comment (ohassine) [inline testservice/src/main/kotlin/com/wire/kalium/testservice/managed/ConversationRepository.kt:132]: This change is from develop, we don't need it in RC

File: testservice/src/main/kotlin/com/wire/kalium/testservice/managed/InstanceService.kt [hunk 8]
```diff
@@ -250,6 +250,8 @@ class InstanceService(
 
                             return@runBlocking instance
                         }
+                        is RegisterClientResult.E2EICertificateRequired ->
+                            throw WebApplicationException("Instance $instanceId: Client registration blocked by e2ei")
```
Comment (ohassine) [inline testservice/src/main/kotlin/com/wire/kalium/testservice/managed/InstanceService.kt:254]: same changes from develop

File: testservice/src/main/kotlin/com/wire/kalium/testservice/models/SendButtonActionConfirmationRequest.kt [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.testservice.models
+
+data class SendButtonActionConfirmationRequest(
```
Comment (ohassine) [inline testservice/src/main/kotlin/com/wire/kalium/testservice/models/SendButtonActionConfirmationRequest.kt:20]: also this

File: testservice/src/main/kotlin/com/wire/kalium/testservice/models/SendButtonActionRequest.kt [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.testservice.models
+
+data class SendButtonActionRequest(
```
Comment (ohassine) [inline testservice/src/main/kotlin/com/wire/kalium/testservice/models/SendButtonActionRequest.kt:20]: and this

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt (hunk 4 above)
Comment (mchenani) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt:1384]: @boris has a fix later we do.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/FinalizeMLSClientAfterE2EIEnrollment.kt (hunk 6 above)
Comment (mchenani) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/FinalizeMLSClientAfterE2EIEnrollment.kt:24]: will take care of this later

**Review by ohassine — APPROVED** [review]

**Pull request merged**